include/circt/Dialect/Seq/SeqOps.td

```diff
@@ -254,3 +259,162 @@ def ClockGateOp : SeqOp<"clock_gate", [Pure]> {
     $input `,` $enable (`,` $test_enable^)? attr-dict
   }];
 }
+
+//===----------------------------------------------------------------------===//
+// FIRRTL-flavored memory
+//===----------------------------------------------------------------------===//
+
+def FirMemOp : SeqOp<"firmem", [
+  MemoryEffects<[MemAlloc]>,
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+]> {
+  let summary = "A FIRRTL-flavored memory";
+  let description = [{
+    The `seq.firmem` op represents memories lowered from the FIRRTL dialect. It
+    is used to capture some of the peculiarities of what FIRRTL expects from
+    memories, while still representing them at the HW dialect level.
+
+    A `seq.firmem` declares the memory and captures the memory-level parameters
+    such as width and depth or how read/write collisions are resolved. The read,
+    write, and read-write ports are expressed as separate operations that take
+    the declared memory as an operand.
+  }];
+
+  let arguments = (ins
+    I32Attr:$readLatency,
+    I32Attr:$writeLatency,
+    RUWAttr:$ruw,
+    WUWAttr:$wuw,
+    OptionalAttr<StrAttr>:$name,
+    OptionalAttr<SymbolNameAttr>:$inner_sym,
+    OptionalAttr<FirMemInitAttr>:$init,
+    OptionalAttr<StrAttr>:$prefix,
+    OptionalAttr<AnyAttr>:$output_file
+  );
+  let results = (outs FirMemType:$memory);
+
+  let assemblyFormat = [{
+    (`sym` $inner_sym^)? `` custom<ImplicitSSAName>($name)
+    $readLatency `,` $writeLatency `,` $ruw `,` $wuw
+    attr-dict `:` type($memory)
+  }];
+}
+
+class AddressMatchesFirMem<string memoryValue, string addressValue> :
+  TypesMatchWith<"address type should match clog2 of memory depth",
+    memoryValue, addressValue, [{
+      IntegerType::get(
+        $_self.getContext(),
+        std::max(1U, llvm::Log2_64_Ceil(cast<FirMemType>($_self).getDepth())))
+    }]>;
+
+class DataMatchesFirMem<string memoryValue, string dataValue> :
+  TypesMatchWith<"data type should match memory width",
+    memoryValue, dataValue, [{
+      IntegerType::get(
+        $_self.getContext(),
+        std::max(1U, cast<FirMemType>($_self).getWidth()))
+    }]>;
+
+def FirMemReadOp : SeqOp<"firmem.read_port", [
```

Comment:
I'm still unsure if make-a-port ops are better than the mem producing the ports.

Reply:
From what I've seen in the Arc dialect which uses this pattern, passes get a lot easier to write if you don't need to do this weird split of multiple ports forming logical groups of operands and results, and then co-iterating the operands and results. Gets especially weird if some ports are optional. With a separate op it's easier to split the port structure into more manageable pieces, and we get port DCE for free.